File: adapters/maze-service/src/main/java/sk/slsp/workshop/mazesolver/adapters/mazeservice/adapter/MazeServiceAdapter.java
package sk.slsp.workshop.mazesolver.adapters.mazeservice.adapter;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import sk.slsp.workshop.mazesolver.Directions;
import sk.slsp.workshop.mazesolver.adapters.mazeservice.port.GoPort;
import sk.slsp.workshop.mazesolver.adapters.mazeservice.port.IsExitPort;
import sk.slsp.workshop.mazesolver.adapters.mazeservice.port.WhereCanIGoPort;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class MazeServiceAdapter implements GoPort, IsExitPort, WhereCanIGoPort {
	
	private final Logger logger = LoggerFactory.getLogger(MazeServiceAdapter.class);
	
	private Integer posX = 1;
	
	private Integer posY = 0;
	
	private final Character[][] maze = {
		{'X', 's', 'X', 'X', 'X'},
		{'X', ' ', 'X', ' ', 'X'},
		{'X', ' ', ' ', ' ', 'X'},
		{'X', ' ', 'X', ' ', 'X'},
		{'X', 'X', 'X', 'e', 'X'}
	};
	
	
	@Override
	public void go(Directions direction) {
		if (direction == Directions.UP) {
			goUp();
		} else if (direction == Directions.DOWN) {
			goDown();
		} else if (direction == Directions.LEFT) {
			goLeft();
		} else if (direction == Directions.RIGHT) {
			goRight();
		}
	}
	
	
	@Override
	public Boolean isExit() {
		return maze[posY][posX] == 'e';
	}
	
	
	@Override
	public List<Directions> whereCanIGo() {
		List<Directions> res = new ArrayList<>();
		
		if (canGoUp()) {
			res.add(Directions.UP);
		}
		if (canGoDown()) {
			res.add(Directions.DOWN);
		}
		if (canGoLeft()) {
			res.add(Directions.LEFT);
		}
		if (canGoRight()) {
			res.add(Directions.RIGHT);
		}
		
		return res;
	}
	
	
	private void goUp() {
		if (canGoUp()) {
			posY -= 1;
		}
		logger.debug(String.format("going up; pos: [%d,%d]", posX, posY));
	}
	
	
	private void goDown() {
		if (canGoDown()) {
			posY += 1;
		}
		logger.debug(String.format("going down; pos: [%d,%d]", posX, posY));
	}
	
	
	private void goLeft() {
		if (canGoLeft()) {
			posX -= 1;
		}
		logger.debug(String.format("going left; pos: [%d,%d]", posX, posY));
	}
	
	
	private void goRight() {
		if (canGoRight()) {
			posX += 1;
		}
		logger.debug(String.format("going right; pos: [%d,%d]", posX, posY));
	}
	
	
	private Boolean canGoUp() {
		return posY > 0 && Arrays.asList('s', 'e', ' ').contains(maze[posY - 1][posX]);
	}
	
	
	private Boolean canGoDown() {
		return posY < (maze.length - 1) && Arrays.asList('s', 'e', ' ').contains(maze[posY + 1][posX]);
	}
	
	
	private Boolean canGoLeft() {
		return posX > 0 && Arrays.asList('s', 'e', ' ').contains(maze[posY][posX - 1]);
	}
	
	
	private Boolean canGoRight() {
		return posX < (maze[0].length - 1) && Arrays.asList('s', 'e', ' ').contains(maze[posY][posX + 1]);
	}
}

File: adapters/maze-service/src/main/kotlin/sk/slsp/workshop/mazesolver/adapters/mazeservice/adapter/MazeServiceAdapter.kt
package sk.slsp.workshop.mazesolver.adapters.mazeservice.adapter

import mu.KotlinLogging
import sk.slsp.workshop.mazesolver.Directions
import sk.slsp.workshop.mazesolver.adapters.mazeservice.port.GoPort
import sk.slsp.workshop.mazesolver.adapters.mazeservice.port.IsExitPort
import sk.slsp.workshop.mazesolver.adapters.mazeservice.port.WhereCanIGoPort

class MazeServiceAdapter:
	WhereCanIGoPort,
	GoPort,
	IsExitPort
{

	private val logger = KotlinLogging.logger {}

	override fun whereCanIGo(): Array<Directions> {
		val res: MutableList<Directions> = ArrayList()

		if (canGoUp()) res.add(Directions.UP)
		if (canGoDown()) res.add(Directions.DOWN)
		if (canGoLeft()) res.add(Directions.LEFT)
		if (canGoRight()) res.add(Directions.RIGHT)

		return res.toTypedArray()
	}

	override fun isExit() = maze[posY][posX] == 'e'

	override fun go(direction: Directions) = when(direction) {
		Directions.UP -> goUp()
		Directions.DOWN -> goDown()
		Directions.LEFT -> goLeft()
		Directions.RIGHT -> goRight()
	}

	private val maze = arrayOf(
		charArrayOf('X', 's', 'X', 'X', 'X'),
		charArrayOf('X', ' ', 'X', ' ', 'X'),
		charArrayOf('X', ' ', ' ', ' ', 'X'),
		charArrayOf('X', ' ', 'X', ' ', 'X'),
		charArrayOf('X', 'X', 'X', 'e', 'X')
	)

	private var posX = 1

	private var posY = 0

	private fun goUp() {
		if (canGoUp()) posY -= 1
		logger.debug("going up; pos: [$posX,$posY]")
	}

	private fun goDown() {
		if (canGoDown()) posY += 1
		logger.debug("going down; pos: [$posX,$posY]")
	}

	private fun goLeft() {
		if (canGoLeft()) posX -= 1
		logger.debug("going left; pos: [$posX,$posY]")
	}

	private fun goRight() {
		if (canGoRight()) posX += 1
		logger.debug("going right; pos: [$posX,$posY]")
	}

	private fun canGoUp() =
		posY > 0 && listOf('s', 'e', ' ').contains(maze[posY-1][posX])

	private fun canGoDown() =
		posY < (maze.size-1) && listOf('s', 'e', ' ').contains(maze[posY+1][posX])

	private fun canGoLeft() =
		posX > 0 && listOf('s', 'e', ' ').contains(maze[posY][posX-1])

	private fun canGoRight() =
		posX < (maze[0].size-1) && listOf('s', 'e', ' ').contains(maze[posY][posX+1])


}
